tests/cxx/explicit_instantiation2.cc

```diff
@@ -53,7 +53,6 @@ FullUseArg<
     // IWYU: Template is...*explicit_instantiation-template.h
     t3; // 3
 
-// IWYU: Template is...*template_short.h.*for explicit instantiation
```

Comment:
This one makes me a little nervous, since there's a an explicit specialization declaration for `Template<short>` in `...template_short.h`. But there's also an explicit specialization definition of that same specialization in this file, so I'm not sure why the remote decl would be crucial.

Reply:
I don't think that reporting of explicit instantiation of `Template<short>` is actually needed for a fwd-decl use inside `FwdDeclUseArg`.
> But there's also an explicit specialization definition of that same specialization in this file, so I'm not sure why the remote decl would be crucial.

Yes, that seems strange, and all this test (maybe after line 42) could become broken after some explicit instantiation analysis improval.

Reply:
I think [this section](https://github.com/include-what-you-use/include-what-you-use/blob/master/iwyu.cc#L3268) is what does it.

If `CanForwardDeclareType` works better now, the explicit instantiation won't be reporteed, I guess.

Moreover, that `ReportExplicitInstantiation` implementation looks naive -- I think we should do something like:
* collect all redecls that are explicit instantiations and not provided by template
* if one exists in same file, use it
* if one exists in another file, and is a declaration, use it (on the assumption that an `extern template ...` declaration is cheaper to process than a full template instantiation)
* otherwise, use an arbitrary one